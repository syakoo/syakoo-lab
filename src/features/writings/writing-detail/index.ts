export { WritingDetail } from "./writing-detail";
