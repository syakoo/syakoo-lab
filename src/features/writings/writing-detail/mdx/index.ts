import { Big } from "./big";
import { BookView } from "./book-view";
import { Chat } from "./chat";
import { Choices, Choice } from "./choices";
import { CodeSandbox } from "./code-sandbox";
import { Figure } from "./figure";
import * as GeometryComponents from "./geometry";
import { Image } from "./image";
import { LinkCard } from "./link-card";
import { Note } from "./note";
import { Scrap } from "./scrap";
import { SectionTitle, SubSectionTitle } from "./section-title";

export const mdxComponents = {
  Big,
  BookView,
  Chat,
  Note,
  CodeSandbox,
  ...GeometryComponents,
  LinkCard,
  Choice,
  Choices,
  Scrap,
  Image,
  SectionTitle,
  SubSectionTitle,
  Figure,
};
