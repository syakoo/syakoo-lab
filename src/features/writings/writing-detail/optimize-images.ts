"use server";

import { readFileSync } from "fs";

import { glob } from "glob";
import type { Sharp } from "sharp";
import sharp from "sharp";

export const optimizeImages = async () => {
  const imagePaths = await glob("**/public/img/writings/**/*.{png,jpg,jpeg}");

  for (const imagePath of imagePaths) {
    const image = sharp(readFileSync(imagePath));
    const optimizedImage = await optimizeImage(image);
    if (optimizedImage) {
      await optimizedImage.toFile(imagePath);
    }
  }
};

const MAX_WIDTH = 2048;

const optimizeImage = async (image: Sharp): Promise<Sharp | undefined> => {
  const metadata = await image.metadata();
  if (!metadata.width || metadata.width <= MAX_WIDTH) return;

  return image.resize({ width: MAX_WIDTH });
};
