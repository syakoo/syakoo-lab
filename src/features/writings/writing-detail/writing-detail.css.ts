import { globalStyle, style } from "@vanilla-extract/css";

import { theme } from "@/design-system/theme.css";
import { headerHeightPx } from "@/features/layout/header-footer-template/header/header.css";

// 縦配置のメディア条件式
const columnLayoutScreen = `screen and (max-width: 800px)` as const;

export const writingDetailStyles = {
  root: style({
    display: "grid",
    maxWidth: 1500,
    margin: "auto",
    justifyContent: "center",
    gridTemplateColumns: `1fr minmax(${theme.size[50]}, ${theme.size[100]}) minmax(${theme.size[50]}, 1fr)`,
    gridTemplateRows: "auto auto",
    gridTemplateAreas: `"l header r"
                        "l content aside"`,

    "@media": {
      [columnLayoutScreen]: {
        gridTemplateColumns: "100%",
        gridTemplateRows: "auto auto",
        gridTemplateAreas: `"header"
                            "content"`,
      },
    },
  }),
  headerWrapper: style({
    gridArea: "header",
    padding: theme.space[200],
    paddingBlock: theme.space[400],
  }),
  contentWrapper: style({
    gridArea: "content",
    padding: theme.space[200],
  }),
  asideWrapper: style({
    gridArea: "aside",
    padding: theme.space[200],
    display: "flex",
    flexDirection: "column",
    gap: theme.space[200],
    height: "100%",

    "@media": {
      [columnLayoutScreen]: {
        display: "none",
      },
    },
  }),
  stickyContainer: style({
    position: "sticky",
    top: `calc(${theme.space[200]} + ${headerHeightPx}px)`,
  }),
};

// 以下 MDX のスタイル
/**
 * MDX コンテナのクラス名
 */
const cn = writingDetailStyles.contentWrapper;

// NOTE: 一つ目の要素は上の空白を消す
globalStyle(`${cn} > *:first-child`, {
  marginTop: "0 !important",
});

// p
globalStyle(`${cn} p`, {
  lineHeight: 2.1,
});
globalStyle(`${cn} p + p`, {
  marginTop: theme.space[300],
});

// ul
globalStyle(`${cn} ul`, {
  listStyleType: "disc",
  marginLeft: theme.space[300],
  marginBlock: theme.space[200],
});
globalStyle(`${cn} ul li::marker`, {
  color: theme.color.text.tertiary,
});
// ol
globalStyle(`${cn} ol`, {
  marginLeft: theme.space[300],
  marginBlock: theme.space[200],
});
globalStyle(`${cn} ol li::marker`, {
  color: theme.color.text.secondary,
});

// code
globalStyle(`${cn} pre`, {
  display: "flex",
  marginBlock: theme.space[200],
  marginInline: `calc(-1 * ${theme.space[200]})`,
  overflowX: "auto",
  color: theme.color.code.text,
  backgroundColor: theme.color.code.background,
  borderRadius: theme.radius[100],

  "@media": {
    [columnLayoutScreen]: {
      borderRadius: 0,
    },
  },
});
globalStyle(`${cn} pre code`, {
  display: "block",
  padding: theme.space[200],
  fontSize: theme.fontSize[75],
  lineHeight: 1.5,
});
// インラインの code
globalStyle(`${cn} code`, {
  color: theme.color.code.text,
  backgroundColor: theme.color.code.background,
  fontSize: theme.fontSize[75],
  padding: "0.3em",
  marginInline: theme.space[25],
  borderRadius: theme.radius[50],
  fontFamily: theme.fontFamily.code,
});

// blockquote
globalStyle(`${cn} blockquote`, {
  padding: `0 ${theme.space[100]}`,
  margin: `${theme.space[300]} 0`,
  color: theme.color.text.secondary,
  borderLeft: `5px solid ${theme.color.text.tertiary}`,
});

globalStyle(`${cn} hr`, {
  marginBlock: theme.space[100],
  borderColor: theme.color.palette.gray[200],
});

// strong
globalStyle(`${cn} strong`, {
  fontWeight: 700,
});

// katex
globalStyle(`${cn} .katex-display`, {
  marginBlock: theme.space[200],
  overflowX: "auto",
  overflowY: "hidden",
});
// mermaid
globalStyle(`${cn} .mermaid`, {
  display: "flex",
  justifyContent: "center",
});
// X (twitter)
globalStyle(`${cn} .twitter-tweet`, {
  marginInline: "auto",
});
