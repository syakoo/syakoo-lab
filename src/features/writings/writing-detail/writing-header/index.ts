export { WritingHeader } from "./writing-header";
