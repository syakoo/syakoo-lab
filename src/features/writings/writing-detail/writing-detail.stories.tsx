import type { Meta, StoryObj } from "@storybook/react";

import { WritingDetailView } from "./writing-detail.view";

const meta = {
  component: WritingDetailView,
  parameters: { testLevel: "snapshot" },
} satisfies Meta<typeof WritingDetailView>;

export default meta;

type Story = StoryObj<typeof meta>;
export const Sample: Story = {
  args: {
    writing: {
      head: {
        id: "sample-writing-meta-id1",
        title: "Sample Article",
        type: "article",
        tags: ["雑記", "Sample"],
        published: "2023-10-09",
      },
      body: {
        type: "serialized",
        data: `
`,
      },
    },
  },
};
