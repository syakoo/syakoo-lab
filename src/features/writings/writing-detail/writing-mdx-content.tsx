"use client";

import type { SerializedWriting } from "@/entities/writing/models/writing";
import { useMermaid } from "@/features/mdx/plugins/mermaid/use-mermaid";
import { useTwitter } from "@/features/mdx/plugins/twitter/use-twitter";
import { resolveMDXContent } from "@/features/mdx/resolver";
import { Link } from "@/shared/design-system/ui";

import { mdxComponents } from "./mdx";

type WritingMdxContentProps = {
  body: SerializedWriting["body"];
};

const components = {
  ...mdxComponents,
  a: (props: Parameters<typeof Link>[0]) => <Link colored {...props} />,
};

export const WritingMdxContent: React.FC<WritingMdxContentProps> = ({
  body,
}) => {
  useMermaid();
  useTwitter();
  const MDXComponent = resolveMDXContent(body).data;

  return <MDXComponent components={components} />;
};
