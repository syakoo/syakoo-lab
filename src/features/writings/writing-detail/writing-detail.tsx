import { differenceInYears } from "date-fns";

import type { SerializedWriting } from "@/features/writings/_models/types";
import { Spacer } from "@/shared/design-system/layout";
import { FadeIn } from "@/shared/design-system/ui";

import { Note } from "./mdx/note";
import { TOC } from "./toc";
import { writingDetailStyles } from "./writing-detail.css";
import { WritingHeader } from "./writing-header";
import { WritingMdxContent } from "./writing-mdx-content";
import { WritingTypeDescription } from "./writing-type-description";

type WritingDetailProps = {
  writing: SerializedWriting;
};

export const WritingDetail: React.FC<WritingDetailProps> = ({ writing }) => {
  const yearsSinceLastUpdate = (() => {
    const writingUpdated = writing.head.updated ?? writing.head.published;
    return differenceInYears(new Date(), new Date(writingUpdated));
  })();

  return (
    <FadeIn>
      <article className={writingDetailStyles.root}>
        <div className={writingDetailStyles.headerWrapper}>
          <WritingHeader head={writing.head} />
        </div>
        <div className={writingDetailStyles.contentWrapper}>
          {/* NOTE: Diary の記事は古くても良い */}
          {writing.head.type !== "diary" && yearsSinceLastUpdate >= 1 ? (
            <Note variant="warn">
              この記事は最終更新日から{yearsSinceLastUpdate}年経過しています。
              内容が古い可能性があります。
            </Note>
          ) : null}
          <WritingMdxContent body={writing.body} />
          <Spacer y="400" />
          <hr />
          <Spacer y="100" />
          <aside>
            <WritingTypeDescription type={writing.head.type} />
          </aside>
        </div>
        <aside className={writingDetailStyles.asideWrapper}>
          <div className={writingDetailStyles.stickyContainer}>
            <TOC />
          </div>
        </aside>
      </article>
    </FadeIn>
  );
};
