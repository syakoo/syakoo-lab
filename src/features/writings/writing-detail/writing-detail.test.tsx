import { testStories } from "@/shared/test-utils/test-stories";

import * as stories from "./writing-detail.stories";

describe("WritingDetailView", () => {
  testStories(stories);
});
