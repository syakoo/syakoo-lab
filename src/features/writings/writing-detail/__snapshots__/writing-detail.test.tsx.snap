// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WritingDetail スナップショット Sample 1`] = `
<body>
  <div>
    <div
      class="fade-in_fadeInStyle__mal1bb1"
      style="animation-duration: 0.5s; animation-delay: 0s;"
    >
      <article
        class="writing-detail_writingDetailStyles_root__jsewnw0"
      >
        <div
          class="writing-detail_writingDetailStyles_headerWrapper__jsewnw1"
        >
          <header>
            <h1
              class="text_textStyle__xbx8xue text_textStyle_color_currentColor__xbx8xui text_textStyle_variants_color_currentColor__xbx8xu3 text_textStyle_size_600__xbx8xuq text_textStyle_variants_size_600__xbx8xub text_textStyle_weight_bold__xbx8xut"
            >
              Sample Article
            </h1>
            <div
              aria-hidden="true"
              class="spacer__1k4u8xae spacer_spacerStyle_y_100__1k4u8xao spacer_spacerStyle_variants_y_100__1k4u8xa9"
            />
            <div
              class="flex_flexStyle__12xne02i flex_flexStyle_align_stretch__12xne02u flex_flexStyle_variants_align_stretch__12xne029 flex_flexStyle_direction_row__12xne02j flex_flexStyle_gap_100__12xne02y flex_flexStyle_variants_gap_100__12xne02d flex_flexStyle_justify_flexStart__12xne02l flex_flexStyle_variants_justify_flexStart__12xne020"
            >
              <div
                class="flex_flexStyle__12xne02i flex_flexStyle_align_center__12xne02s flex_flexStyle_variants_align_center__12xne027 flex_flexStyle_direction_row__12xne02j flex_flexStyle_gap_50__12xne02x flex_flexStyle_variants_gap_50__12xne02c flex_flexStyle_justify_flexStart__12xne02l flex_flexStyle_variants_justify_flexStart__12xne020"
              >
                <svg
                  fill="none"
                  stroke="var(--color-text-secondary__13jk9q1d)"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.5 14.25v-2.625a3.375 3.375 0 00-3.375-3.375h-1.5A1.125 1.125 0 0113.5 7.125v-1.5a3.375 3.375 0 00-3.375-3.375H8.25m0 12.75h7.5m-7.5 3H12M10.5 2.25H5.625c-.621 0-1.125.504-1.125 1.125v17.25c0 .621.504 1.125 1.125 1.125h12.75c.621 0 1.125-.504 1.125-1.125V11.25a9 9 0 00-9-9z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                <p
                  class="text_textStyle__xbx8xue text_textStyle_color_secondary__xbx8xug text_textStyle_variants_color_secondary__xbx8xu1 text_textStyle_size_75__xbx8xuk text_textStyle_variants_size_75__xbx8xu5 text_textStyle_weight_normal__xbx8xuu"
                >
                  Article
                </p>
              </div>
              <p
                class="text_textStyle__xbx8xue text_textStyle_color_secondary__xbx8xug text_textStyle_variants_color_secondary__xbx8xu1 text_textStyle_size_75__xbx8xuk text_textStyle_variants_size_75__xbx8xu5 text_textStyle_weight_normal__xbx8xuu"
              >
                #
                雑記
              </p>
              <p
                class="text_textStyle__xbx8xue text_textStyle_color_secondary__xbx8xug text_textStyle_variants_color_secondary__xbx8xu1 text_textStyle_size_75__xbx8xuk text_textStyle_variants_size_75__xbx8xu5 text_textStyle_weight_normal__xbx8xuu"
              >
                #
                Sample
              </p>
            </div>
            <div>
              <a
                class="link_linkStyle__jc9bkl0"
                href="/"
              >
                <span
                  class="text_textStyle__xbx8xue text_textStyle_color_secondary__xbx8xug text_textStyle_variants_color_secondary__xbx8xu1 text_textStyle_size_75__xbx8xuk text_textStyle_variants_size_75__xbx8xu5 text_textStyle_weight_normal__xbx8xuu"
                >
                  syakoo
                </span>
              </a>
              <span
                class="text_textStyle__xbx8xue text_textStyle_color_secondary__xbx8xug text_textStyle_variants_color_secondary__xbx8xu1 text_textStyle_size_75__xbx8xuk text_textStyle_variants_size_75__xbx8xu5 text_textStyle_weight_normal__xbx8xuu"
              >
                ・
                <time
                  datetime="2023-10-09"
                >
                  2023-10-09
                </time>
              </span>
            </div>
          </header>
        </div>
        <div
          class="writing-detail_writingDetailStyles_contentWrapper__jsewnw2"
        >
          <aside
            class="note_noteStyles_root__1sjosdw0 note_noteStyles_root_variant_warn__1sjosdw2"
          >
            <div
              class="note_noteStyles_iconWrapper__1sjosdw3 note_noteStyles_iconWrapper_variant_warn__1sjosdw5"
            >
              <svg
                fill="none"
                stroke="currentColor"
                stroke-width="1.5"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 9v3.75m-9.303 3.376c-.866 1.5.217 3.374 1.948 3.374h14.71c1.73 0 2.813-1.874 1.948-3.374L13.949 3.378c-.866-1.5-3.032-1.5-3.898 0L2.697 16.126zM12 15.75h.007v.008H12v-.008z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                />
              </svg>
            </div>
            <div>
              この記事は古い内容である可能性があります。
            </div>
          </aside>
          <p>
            Sample MDX Component
          </p>
          <h2>
            Sample Title
          </h2>
          <div
            aria-hidden="true"
            class="spacer__1k4u8xae spacer_spacerStyle_y_400__1k4u8xar spacer_spacerStyle_variants_y_400__1k4u8xac"
          />
          <hr />
          <div
            aria-hidden="true"
            class="spacer__1k4u8xae spacer_spacerStyle_y_100__1k4u8xao spacer_spacerStyle_variants_y_100__1k4u8xa9"
          />
          <aside>
            <section
              class="flex_flexStyle__12xne02i flex_flexStyle_align_stretch__12xne02u flex_flexStyle_variants_align_stretch__12xne029 flex_flexStyle_direction_column__12xne02k flex_flexStyle_gap_50__12xne02x flex_flexStyle_variants_gap_50__12xne02c flex_flexStyle_justify_flexStart__12xne02l flex_flexStyle_variants_justify_flexStart__12xne020"
            >
              <div
                class="flex_flexStyle__12xne02i flex_flexStyle_align_center__12xne02s flex_flexStyle_variants_align_center__12xne027 flex_flexStyle_direction_row__12xne02j flex_flexStyle_gap_50__12xne02x flex_flexStyle_variants_gap_50__12xne02c flex_flexStyle_justify_flexStart__12xne02l flex_flexStyle_variants_justify_flexStart__12xne020"
              >
                <svg
                  fill="none"
                  stroke="currentColor"
                  stroke-width="1.5"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19.5 14.25v-2.625a3.375 3.375 0 00-3.375-3.375h-1.5A1.125 1.125 0 0113.5 7.125v-1.5a3.375 3.375 0 00-3.375-3.375H8.25m0 12.75h7.5m-7.5 3H12M10.5 2.25H5.625c-.621 0-1.125.504-1.125 1.125v17.25c0 .621.504 1.125 1.125 1.125h12.75c.621 0 1.125-.504 1.125-1.125V11.25a9 9 0 00-9-9z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                <p
                  class="text_textStyle__xbx8xue text_textStyle_color_currentColor__xbx8xui text_textStyle_variants_color_currentColor__xbx8xu3 text_textStyle_size_100__xbx8xul text_textStyle_variants_size_100__xbx8xu6 text_textStyle_weight_bold__xbx8xut"
                >
                  Article
                </p>
              </div>
              <div>
                <p
                  class="text_textStyle__xbx8xue text_textStyle_color_secondary__xbx8xug text_textStyle_variants_color_secondary__xbx8xu1 text_textStyle_size_75__xbx8xuk text_textStyle_variants_size_75__xbx8xu5 text_textStyle_weight_normal__xbx8xuu"
                >
                  Article は気合いの入った文書を残しています。
                </p>
              </div>
            </section>
          </aside>
        </div>
        <aside
          class="writing-detail_writingDetailStyles_asideWrapper__jsewnw3"
        >
          <div
            class="writing-detail_writingDetailStyles_stickyContainer__jsewnw4"
          >
            <nav>
              <p
                class="text_textStyle__xbx8xue text_textStyle_color_currentColor__xbx8xui text_textStyle_variants_color_currentColor__xbx8xu3 text_textStyle_size_100__xbx8xul text_textStyle_variants_size_100__xbx8xu6 text_textStyle_weight_bold__xbx8xut"
              >
                目次
              </p>
              <ul
                class="toc_tocStyles_list__1yz9ns40"
              >
                <li
                  class="toc_tocStyles_item__1yz9ns41"
                >
                  <a
                    class="link_linkStyle__jc9bkl0 link_linkStyle_noHovered_true__jc9bkl3 link_linkStyle_display_block__jc9bkl4"
                    href="#"
                  >
                    Sample Title
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </aside>
      </article>
    </div>
  </div>
  <script
    src="https://platform.twitter.com/widgets.js"
  />
</body>
`;
